### PR TITLE
Update NVBench fixture to use new hooks, fix pinned memory segfault.

### PR DESCRIPTION
**Pull request by alliepiper**
## Description

NVBench recently exposed new hooks for modifying its `main` implementation. Updated cudf to use these.

Also noticed that the host pinned-pool memory resource option caused the test to segfault, since the function-scope static holding the pool outlived the CUDA context. Refactored the fixture a bit to ensure that the pool is destroyed before the context.

Note that this currently overrides the rapids-cmake version for NVBench. Rapids-cmake should be updated and the override removed before this is merged (ping @robertmaynard).

cc: @jrhemstad @davidwendt

### CODE REVIEW

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (bdice) [inline cpp/cmake/thirdparty/patches/nvbench_override.json:6]: This tag should be updated in https://github.com/rapidsai/rapids-cmake/blob/8b1a1e0e2302ec5a6cfeed762c4f281268e7adca/rapids-cmake/cpm/versions.json#L84 rather than here.

Comment (jrhemstad) [reply cpp/cmake/thirdparty/patches/nvbench_override.json:6]: @bdice what's the process for making PR's that depend on rapids-cmake changes? 

I don't suppose there's a way to make a cudf PR target the rapids-cmake PR? Or do we need to wait for the rapids-cmake changes to get merged first?

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (davidwendt) [inline cpp/benchmarks/fixture/nvbench_main.cpp:26]: I would be more comfortable with a less generic name since this is in the global namespace.
Something like `cudf_benchmark_arg_handler` or maybe a namespace like `cudf::benchmark_arg_handler`.

File: cpp/benchmarks/fixture/nvbench_fixture.hpp
Comment (pmattione-nvidia) [inline cpp/benchmarks/fixture/nvbench_fixture.hpp:96]: Should this function guard against being called twice?

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (davidwendt) [inline cpp/benchmarks/fixture/nvbench_main.cpp:38]:
```suggestion
```

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (davidwendt) [reply cpp/benchmarks/fixture/nvbench_main.cpp:30]: Cool. Could you write/post the snippet here?
It needs to recognize 2 strings and if either is found remove that entry plus the very next entry.

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (pmattione-nvidia) [inline cpp/benchmarks/fixture/nvbench_main.cpp:43]:
```suggestion
  bool remove_next = false;
  auto const args_end = args.end();
  args.erase(std::remove_if(args.begin(), args_end, [&](const auto& arg){
    if(remove_next) {
      remove_next = false;
      return true;
    }
    if ((arg == cudf::detail::rmm_mode_param) || (arg == cudf::detail::cuio_host_mem_param)) {
      remove_next = true;
      return true;
    }
    return false;
  }), args_end);
```

Comment (davidwendt) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: I'm not sure this implementation is clearer but I'm ok with it.
Stateful lambdas make me cringe a bit. I suppose I'm used to them being parallelizable.

Comment (pmattione-nvidia) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: I'm also ok with an argument that performance isn't as important as clarity here, just a suggestion.

Comment (harrism) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: I doubt it would be faster or clearer. Why not just use cxxopts. We already depend on cxxopts in libcudf.

Comment (davidwendt) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: The point of this function (called by nvbench) is to filter out options from the given vector. 
Actual handling of these parameters is done in a different file (nvbench_fixture.hpp).

Comment (harrism) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: Maybe nvbench should use cxxopts.

Comment (davidwendt) [reply cpp/benchmarks/fixture/nvbench_main.cpp:43]: Perhaps. I would consider that outside the scope of this PR though since that code (in `nvbench_fixture.hpp` -- not the file this comment chain is attached to) is not required to be updated for this change.

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (robertmaynard) [reply cpp/cmake/thirdparty/patches/nvbench_override.json:6]: https://github.com/rapidsai/rapids-cmake/pull/584 is now active to merge these changes upstream.
But once merged it will break cudf benchmarks since our current patch will fail to apply.

So to ensure no breaks, we will need to merge this PR first, merge the rapids-cmake PR, and make a follow-up that removes the `nvbench_override.json` ( and related CMake call ).

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (davidwendt) [inline cpp/benchmarks/fixture/nvbench_main.cpp:38]:
```suggestion
    if (arg == cudf::detail::rmm_mode_param) {
      i++;  // skip the next argument
    } else if (arg == cudf::detail::cuio_host_mem_param) {
      i++;  // skip the next argument
    } else {
```

**Review by davidwendt — APPROVED** [review]

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (bdice) [reply cpp/cmake/thirdparty/patches/nvbench_override.json:6]: We should use the same commit in this PR and the rapids-cmake PR:
```suggestion
      "git_tag": "555d628e9b250868c9da003e4407087ff1982e8e"
```

File: cpp/benchmarks/fixture/nvbench_fixture.hpp
Comment (alliepiper) [reply cpp/benchmarks/fixture/nvbench_fixture.hpp:96]: Yes -- fixed.

File: cpp/benchmarks/fixture/nvbench_main.cpp
Comment (alliepiper) [reply cpp/benchmarks/fixture/nvbench_main.cpp:26]: Good point, I'll make this change.

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (alliepiper) [reply cpp/cmake/thirdparty/patches/nvbench_override.json:6]: Updated the SHA. I'll leave this change active following Rob's suggestion above.

**Comment by alliepiper** [comment]
Thanks for updating this @davidwendt -- I made a couple more changes to address some of the other comments.

**Review by pmattione-nvidia — APPROVED** [review]

File: cpp/cmake/thirdparty/patches/nvbench_override.json
Comment (davidwendt) [reply cpp/cmake/thirdparty/patches/nvbench_override.json]: @robertmaynard I think we are ready to merge this so we can start the merge dance you described above.

Comment (robertmaynard) [reply cpp/cmake/thirdparty/patches/nvbench_override.json]: Sounds great. :+1:

**Review by robertmaynard — APPROVED** [review]

**Pull request merged**
